runtime/interpreter/numerical_value.go

```diff
@@ -0,0 +1,11236 @@
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationPlus,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	// INT32-C
+	if (o > 0) && (v > (math.MaxInt8 - o)) {
+		panic(OverflowError{locationRange})
+	} else if (o < 0) && (v < (math.MinInt8 - o)) {
+		panic(UnderflowError{locationRange})
+	}
+
+	valueGetter := func() int8 {
+		return int8(v + o)
+	}
+
+	return NewInt8Value(interpreter, valueGetter)
+}
+
+func (v Int8Value) SaturatingPlus(interpreter *Interpreter, other NumberValue, locationRange LocationRange) NumberValue {
+	o, ok := other.(Int8Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			FunctionName: sema.NumericTypeSaturatingAddFunctionName,
+			LeftType:     v.StaticType(interpreter),
+			RightType:    other.StaticType(interpreter),
+		})
+	}
+
+	valueGetter := func() int8 {
+		// INT32-C
+		if (o > 0) && (v > (math.MaxInt8 - o)) {
+			return math.MaxInt8
+		} else if (o < 0) && (v < (math.MinInt8 - o)) {
+			return math.MinInt8
+		}
+		return int8(v + o)
+	}
+
+	return NewInt8Value(interpreter, valueGetter)
+}
+
+func (v Int8Value) Minus(interpreter *Interpreter, other NumberValue, locationRange LocationRange) NumberValue {
+	o, ok := other.(Int8Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationMinus,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	// INT32-C
+	if (o > 0) && (v < (math.MinInt8 + o)) {
+		panic(OverflowError{locationRange})
+	} else if (o < 0) && (v > (math.MaxInt8 + o)) {
+		panic(UnderflowError{locationRange})
+	}
+
+	valueGetter := func() int8 {
+		return int8(v - o)
+	}
+
+	return NewInt8Value(interpreter, valueGetter)
+}
+
+func (v Int8Value) SaturatingMinus(interpreter *Interpreter, other NumberValue, locationRange LocationRange) NumberValue {
+	o, ok := other.(Int8Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			FunctionName: sema.NumericTypeSaturatingSubtractFunctionName,
+			LeftType:     v.StaticType(interpreter),
+			RightType:    other.StaticType(interpreter),
+		})
+	}
+
+	valueGetter := func() int8 {
+		// INT32-C
+		if (o > 0) && (v < (math.MinInt8 + o)) {
+			return math.MinInt8
+		} else if (o < 0) && (v > (math.MaxInt8 + o)) {
+			return math.MaxInt8
+		}
+		return int8(v - o)
+	}
+
+	return NewInt8Value(interpreter, valueGetter)
+}
+
+func (v Int8Value) Mod(interpreter *Interpreter, other NumberValue, locationRange LocationRange) NumberValue {
+	o, ok := other.(Int8Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationMod,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	// INT33-C
+	if o == 0 {
+		panic(DivisionByZeroError{locationRange})
+	}
+
+	valueGetter := func() int8 {
+		return int8(v % o)
+	}
+
+	return NewInt8Value(interpreter, valueGetter)
+}
+
+func (v Int8Value) Mul(interpreter *Interpreter, other NumberValue, locationRange LocationRange) NumberValue {
```

Comment:
While modifying all these functions, I noticed that they are all fairly redundant in terms of their content and implementation; mostly we just compare values against type-dependent maximums and minimums, and overflow/underflow as necessary, performing the arithmetic operation otherwise. It might be nice to rewrite these implementations to use generic functions. Does this seem worth it or would we rather keep these as they are for performance?

Reply:
Yes, if generics can be used to reduce code duplication, while not regressing on performance too much, we should try it 👍

Reply:
Nice, I'll look into a separate PR for this next week.